### PR TITLE
Ensure we always callback user's completion block even when cancelled with `SDWebImageErrorCancelled` error code

### PR DESCRIPTION
**Pull request by dreampiggy**
which can guarantee the completed block of downloader/manager is called back

### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: #2690 #1448 #2539 #2589

### Pull Request Description

This PR fix the current behavior, that sometimes, we do not callback the completion block from manager && downloader. The case exist when user call the following code and there are URLs in pending queue:

```objective-c
[SDWebImageDownloader.sharedDownloader cancelAllDownloads];
```

# Reason

Current implementation of `SDWebImageDownloadOperation`, will only callback the block in `URLSession:task:didCompleteWithError:` delegate method. However, if the download operation is just added to the operation queue, did not started. The cancel of this download operation will be silent, without any callback, notification, etc.

This cause many issue which assume the `the manager/downloader completionBlock should be callback`. And code depends this assume will cause issue in this case

# Solution

We can just learn from another popular image framework: [Kingfisher](https://github.com/onevcat/Kingfisher/blob/master/Sources/General/KingfisherError.swift#L53-L56). Which gurantee the callback is trigged either success or failure. When the user cancel the downloading, it will populate a custom error. To let any code which register the block to know this URL is cancelled.

```
  ▿ requestError : 1 element
    ▿ reason : RequestErrorReason
      ▿ taskCancelled : 2 elements
        ▿ task : <SessionDataTask: 0x60000331ca20>
        - token : 0
```

So we just do the nearly same thing. We using the custom error to distinguish this from the `NSURLErrorCancelled` (Which represent the network issue, may not suitable for user cancel)

```objective-c
typedef NS_ERROR_ENUM(SDWebImageErrorDomain, SDWebImageError) {
    ......
    SDWebImageErrorDownloadCancelled = 2002, // The image download operation is cancelled by user.
};
```

# Solved issue

For example, these following case will not cause any issue after this PR:

+ `SDWebImagePrefetcher`, do not need any code changes. Because even user cancel the operation, the callback is here and also cancel the `SDAsyncBlockOperation`. No hung out.
+ Using `dispatch group enter/leave` in completion block to make async -> sync: Currently because of `the completionBlock is not guranteeded to be called`, this may cause unbalanced and crash 
+ User want to distinguish the `Network Error of cancel` with `User code to trigger cancel`, may have different behavior. Previouslly they can not, because the user cancel is silent, no Notification, no callback, nothing.

### CODE REVIEW

**Comment by dreampiggy** [comment]
@zhongwuzw Actually. It's already the case Kingfisher used. So I think we can just do the same thing, to solve potential issue.

**Comment by dreampiggy** [comment]
Added to 5.0.2 milestone. Seems we can check it later.

**Comment by zhongwuzw** [comment]
Come back, we can't ensure `completionBlock` always be called in this PR, we need to do some refactor internal, I don't know wether it's necessary.

I can list some cases:
1. `SDWebImageCombinedOperation` created and added to `runningOperations`, but its `cacheOperation` or `loaderOperation` not created, now cancel happened.
2. `SDWebImageCombinedOperation` and `cacheOperation` created, but `loaderOperation` not created still not be created, now cancel happened.
3. Users sets `defaultImageCache` and `defaultImageLoader` of `SDWebImageManager`, we lose the control to call `completionBlock`, it would affects sharedManager.
......

**Comment by dreampiggy** [comment]
Emm ?

I just means, if there a `Download Operation` in queue, whenever a `cancel`, or task finished (success or failed), it gurantees to callback.

> Case1
Since the download operation is not created, call any cancel at this time does not means anything. It just run normally.

> Case2
The same as case1

> Case3

Using a custom download operation implementation, people can ensure their own gurantees. So it's OK. We support customization, but not means we support any bad written implementation.

**Comment by dreampiggy** [comment]
>  If there a Download Operation in queue, whenever a cancel, or task finished (success or failed), it gurantees to callback.

This can fix #2690, without changing current user's behavior of prefetcher's config. It still using the default downloader with all pre-defined config like HTTP headers.

**Comment by dreampiggy** [comment]
Actually, any version of SDWebImage in 4.0, using `SDWebImagePrefetcher`, can still have the same issue to hang. See the code. It need the completion callback to trigger next request. https://github.com/SDWebImage/SDWebImage/blob/4.4.6/SDWebImage/SDWebImagePrefetcher.m#L83-L86

So , this issue should be solved from the root case. And maybe we need another patch version release for 4.x version.

**Comment by zhongwuzw** [comment]
> > If there a Download Operation in queue, whenever a cancel, or task finished (success or failed), it gurantees to callback.

Nope, this is not guarantee.

> With this guarantees. By combine with `SDWebImageManager`'s existing logic, we can infer another behavior as well.
> 
> > `SDWebImageManage loadImageWithURL:`, can gurantee to callback, whether the query is cancelled, success or failed, whether it hit the cache or download part.
> 
> So, this can fix #2690, without changing current user's behavior of prefetcher's config. It still using the default downloader with all pre-defined config like HTTP headers.

We can't use default shared downloader.

**Comment by zhongwuzw** [comment]
> Emm ?
> 
> I just means, if there a `Download Operation` in queue, whenever a `cancel`, or task finished (success or failed), it gurantees to callback.
> 
> > Case1
> 
> Since the download operation is not created, call any cancel at this time does not means anything. It just run normally. After load success or failed, it callback normally.

Nope, you can check the code, this is not guarantee.

> > Case2
> 
> The same as case1.
> 
> > Case3
> 
> Using a custom download operation implementation, people can ensure their own gurantees. So it's OK. We support customization, but not means we support any bad written implementation.

Shared prefetcher can't use this shared manager and shared downloader. we can't ensure `completionBlock` always be called, also can't require users to ensure it IMO.

**Comment by dreampiggy** [comment]
Could you please tell the detail of this `Nope, you can check the code, this is not guarantee, even download operation is in queue.`.

I see that the logic, only the `NSOperation.start` and `cancelInternal` can cause the earily return without callback. 

And, I think whatever reason, there should be a gurantee as a framework. Could anyone imagine that `URLSession` didn't callback your completion handler in some case and cause your Apps code run unexcepted ?

**Comment by dreampiggy** [comment]
And, I think it's better for us to track and maintain the massive features and PRs. I start a GitHub board. Which track them. Each implmentation should better provide a bug issue or feature request issue, so that we can track them `To Do`, `In Progress`  and finally get it done.

**Comment by dreampiggy** [comment]
Added the bug issue report: #2702

**Comment by zhongwuzw** [comment]
> And, I think it's better for us to track and maintain the massive features and PRs. I start a GitHub board. Which track them. Each implmentation should better provide a bug issue or feature request issue, so that we can track them `To Do`, `In Progress` and finally get it done.

@dreampiggy  This is awesome. 👍

**Comment by zhongwuzw** [comment]
@bpoplauschi I already list some cases in my comments of this PR and https://github.com/SDWebImage/SDWebImage/issues/2702#issuecomment-491333083, what's your opinion? 🤔 

TLDR.
This PR not fix any bugs like description said, but if we agree that when we cancel a **downloading operation** , to call callbacks added before cancel operation happened, we can add this, but needs conform Semver, maybe `6.0`.

**Comment by dreampiggy** [comment]
The https://github.com/SDWebImage/SDWebImage/pull/2692#issuecomment-491574277 is fixed. So this is not a issue for us to prevent user's request for `ensure completion block is called`.

**Comment by mikk22** [comment]
Since 5.1.0 `SDWebImageDelayPlaceholder` option behavior was changed.
In 5.0.6 while you have several operations and `SDWebImageDelayPlaceholder` you see placeholder image after last operation fail.
In 5.1.0 you see placeholder image after first operation was cancelled.

**Comment by dreampiggy** [comment]
@mikk22 This may because of that the change `cancel is a error, not something we need to ignore`.

Please submit a new GitHub issue and some detail use case about your code. A example is better. I'll find the way to solve this.
